[BUG] case when on some nested types can produce a crash

**Describe the bug**
```
Caused by: ai.rapids.cudf.CudfException: cuDF failure at: /home/roberte/src/cudf/cpp/src/reshape/interleave_columns.cu:38: Called `interleave_columns` on none-supported data type.
E                   	at ai.rapids.cudf.ColumnVector.makeList(Native Method)
E                   	at ai.rapids.cudf.ColumnVector.makeList(ColumnVector.java:441)
E                   	at ai.rapids.cudf.ColumnVector.makeList(ColumnVector.java:410)
E                   	at ai.rapids.cudf.ColumnVector.empty(ColumnVector.java:1623)
E                   	at ai.rapids.cudf.ColumnVector.empty(ColumnVector.java:1630)
E                   	at ai.rapids.cudf.Scalar.listFromNull(Scalar.java:357)
E                   	at com.nvidia.spark.rapids.GpuScalar$.from(literals.scala:221)
E                   	at com.nvidia.spark.rapids.GpuScalar.getBase(literals.scala:442)
E                   	at com.nvidia.spark.rapids.GpuConditionalExpression.$anonfun$computeIfElse$3(conditionalExpressions.scala:42)
```

This happens when you do a case when statement and no default value is given.  The types involved are.

```
Array(Struct(key: String(not_null), value: Array(Struct(key: String(not_null), value: String))))
```

This is the same as a

```
Map(String(not_null),Map(String(not_null),String))
```

but we don't support maps on case when yet. I was trying to add them and ran into this.

--- a/docs/supported_ops.md
+++ b/docs/supported_ops.md
@@ -3024,9 +3024,9 @@ are limited.
 <td>S</td>
 <td><b>NS</b></td>
 <td><b>NS</b></td>
-<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, MAP, UDT</em></td>
-<td><b>NS</b></td>
-<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, MAP, UDT</em></td>
+<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, UDT</em></td>
+<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, UDT</em></td>
+<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, UDT</em></td>
 <td><b>NS</b></td>
 </tr>
 <tr>
@@ -3045,9 +3045,9 @@ are limited.
 <td>S</td>
 <td><b>NS</b></td>
 <td><b>NS</b></td>
-<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, MAP, UDT</em></td>
-<td><b>NS</b></td>
-<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, MAP, UDT</em></td>
+<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, UDT</em></td>
+<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, UDT</em></td>
+<td><em>PS<br/>max nested DECIMAL precision of 18;<br/>UTC is only supported TZ for nested TIMESTAMP;<br/>missing nested BINARY, CALENDAR, UDT</em></td>
 <td><b>NS</b></td>
 </tr>
 <tr>

--- a/integration_tests/src/main/python/conditionals_test.py
+++ b/integration_tests/src/main/python/conditionals_test.py
@@ -21,7 +21,7 @@ from pyspark.sql.types import *
 import pyspark.sql.functions as f
 
 all_gens = all_gen + [NullGen()]
-all_nested_gens = array_gens_sample + struct_gens_sample
+all_nested_gens = array_gens_sample + struct_gens_sample + map_gens_sample
 all_nested_gens_nonempty_struct = array_gens_sample + nonempty_struct_gens_sample
 
 # Create dedicated data gens of nested type for 'if' tests here with two exclusions:

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -657,6 +657,14 @@ def _mark_as_lit(data, data_type):
         # but this works for all of them
         dateString = data.strftime("%Y-%m-%d").zfill(10)
         return f.lit(dateString).cast(data_type)
+    elif isinstance(data_type, MapType):
+        assert isinstance(data, dict)
+        # Sadly you cannot create a literal from just a dict/tuple in pyspark
+        col_array = []
+        for k in data:
+            col_array.append(_mark_as_lit(k, data_type.keyType))
+            col_array.append(_mark_as_lit(data[k], data_type.valueType))
+        return f.create_map(*col_array)
     else:
         # lit does not take a data type so we might have to cast it
         return f.lit(data).cast(data_type)

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala
@@ -997,7 +997,7 @@ case class ExprChecksImpl(contexts: Map[ExpressionContext, ContextChecks])
  */
 object CaseWhenCheck extends ExprChecks {
   val check: TypeSig = (TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL_64 +
-    TypeSig.ARRAY + TypeSig.STRUCT).nested()
+    TypeSig.ARRAY + TypeSig.STRUCT + TypeSig.MAP).nested()
 
   val sparkSig: TypeSig = TypeSig.all
 

